core/src/main/java/com/graphhopper/routing/util/MaxSpeedCalculator.java

```diff
@@ -0,0 +1,153 @@
+package com.graphhopper.routing.util;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.graphhopper.routing.ev.*;
+import com.graphhopper.routing.util.parsers.helpers.OSMValueExtractor;
+import com.graphhopper.storage.Graph;
+import de.westnordost.osm_legal_default_speeds.LegalDefaultSpeeds;
+import de.westnordost.osm_legal_default_speeds.RoadType;
+import de.westnordost.osm_legal_default_speeds.RoadTypeFilter;
+import kotlin.jvm.functions.Function0;
+import kotlin.jvm.functions.Function2;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MaxSpeedCalculator {
+
+    public static void fillMaxSpeed(Graph graph,
+                                    EnumEncodedValue<UrbanDensity> urbanDensityEnc,
+                                    EnumEncodedValue<RoadClass> roadClassEnc,
+                                    BooleanEncodedValue roadClassLinkEnc,
+                                    EnumEncodedValue<Country> countryEnumEncodedValue,
+                                    DecimalEncodedValue maxSpeedEnc) {
+
+        SpeedLimitsJson data;
+        try {
+            data = new ObjectMapper().readValue(MaxSpeedCalculator.class.getResource("legal_default_speeds.json"), SpeedLimitsJson.class);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        LegalDefaultSpeeds spLimit = new LegalDefaultSpeeds(data.roadTypesByName, data.speedLimitsByCountryCode);
+
+        List<Map<String, String>> relTags = new ArrayList<>();
+        // You can replace the result of any number of placeholders in a tag filter here (e.g. for name = "urban"),
+        // for example if you have another data source to acquire whether a road is in a built-up area or not.
+        // For those you do not want to replace, simply pass on the result of evaluate as result
+        Function2<String, Function0<Boolean>, Boolean> replacerFunction = (name, eval) -> eval.invoke();
+        AllEdgesIterator iter = graph.getAllEdges();
+        while (iter.next()) {
+            // ISO 3166-1 alpha-2 code optionally concatenated with a ISO 3166-2 code, e.g. "DE", "US" or "BE-VLG"
+            String countryCode = iter.get(countryEnumEncodedValue).getAlpha2();
+            Map<String, String> tags = new HashMap<>();
+            tags.put("rural", iter.get(urbanDensityEnc) == UrbanDensity.RURAL ? "yes" : "no");
+            tags.put("highway", iter.get(roadClassEnc).toString() + (iter.get(roadClassLinkEnc) ? "_link" : ""));
```

Comment:
You shouldn't need to do that (if I understand that piece of code correctly):

See e.g.
https://westnordost.github.io/osm-legal-default-speeds/#tags=highway%3Dmotorway_link&cc=DE

Reply:
What I mean to say is that the filters already include `<something>_link`. E.g. Search for `motorway` in this table:

https://wiki.openstreetmap.org/wiki/Default_speed_limits#Road_types_to_tag_filters

The filter is `highway~motorway|motorway_link`

If for any country, different speed limits exist for motorway ramps / links than for motorways, it would be noted in the table. If e.g. in South Sudan, motorways had a speed limit of 100 and motorway ramps a speed limit of 80, the table would look like this:

<table>
<tr><th rowspan="2">South Sudan</th><td>motorway</td><td>100</td></tr>
<tr><td>motorway link</td><td>80</td></tr>
</table>

and in the road types table, there would be an entry like

<table>
<tr><td>motorway link</td><td><code>highway=motorway_link</code></td></tr>
</table>